### ISSUE TITLE
Invoking content assist on missing class import caused stuck UI thread

### ISSUE THREAD

**Issue by gamerson** [bndtools/bnd#3956]
In a Bnd/gradle workspace, while editing a java file, I typed in some code that needs a class import and when I invoked the quickfix shortcut Cmd+1, the UI thread froze.

Here is the jstack output for the process
https://gist.github.com/gamerson/f7d143c955786741fc0149d8c71bd518

Even though the main thread was reading from a network socket, nettop said that the process had no RX/TX packets so it appears to be doing nothing on the actual network, I could be reading nettop wrong however.

**Comment by bjhargrave** [comment]
This seems to be doing work that should not be done on the main UI thread:

```
	at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(java.base@11.0.2/HttpURLConnection.java:3495)
	at java.io.FilterInputStream.read(java.base@11.0.2/FilterInputStream.java:133)
	at java.io.PushbackInputStream.read(java.base@11.0.2/PushbackInputStream.java:183)
	at java.util.zip.InflaterInputStream.fill(java.base@11.0.2/InflaterInputStream.java:243)
	at java.util.zip.InflaterInputStream.read(java.base@11.0.2/InflaterInputStream.java:159)
	at java.util.zip.ZipInputStream.read(java.base@11.0.2/ZipInputStream.java:195)
	at java.util.zip.ZipInputStream.closeEntry(java.base@11.0.2/ZipInputStream.java:141)
	at java.util.zip.ZipInputStream.getNextEntry(java.base@11.0.2/ZipInputStream.java:119)
	at aQute.lib.zip.JarIndex.buildFromInputStream(JarIndex.java:138)
	at aQute.lib.zip.JarIndex.<init>(JarIndex.java:49)
	at aQute.bnd.build.WorkspaceClassIndex.matchClassNameAgainstResource(WorkspaceClassIndex.java:98)
	at aQute.bnd.build.WorkspaceClassIndex.lambda$search$1(WorkspaceClassIndex.java:83)
	at aQute.bnd.build.WorkspaceClassIndex$$Lambda$1716/0x00000008015da040.accept(Unknown Source)
	at java.util.Optional.ifPresent(java.base@11.0.2/Optional.java:183)
	at aQute.bnd.build.WorkspaceClassIndex.search(WorkspaceClassIndex.java:82)
	at aQute.bnd.build.Workspace.search(Workspace.java:1437)
	at org.bndtools.core.editors.quickfix.BuildpathQuickFixProcessor.getCorrections(BuildpathQuickFixProcessor.java:92)
```

Also, why is it trying to index something at an http URL? I would think we would only index local files.

**Comment by bjhargrave** [comment]
The code is trying to index a resource. It gets the URI from the content capability and then tries to read the content to index the jar. For some reason your connection hung.

I am wondering if the indexing code here should only index local files as indexing remote files like this can be problematic...

**Comment by bjhargrave** [comment]
I have a fix in a PR to always use ZipFile for local files instead of ZipInputStream. But your case says there is a remote http url in use.

**Comment by kriegfrj** [comment]
@gamerson noted that this occurred when he was resolving against 4 big p2 repositories. The P2Repository has to download all the bundles to index them, and it could be that this has been abstracted away I'm guessing that's probably what happened.

> The code is trying to index a resource. It gets the URI from the content capability and then tries to read the content to index the jar. For some reason your connection hung.
> 
> I am wondering if the indexing code here should only index local files as indexing remote files like this can be problematic...

II like the idea of having the bundles of you remote repos available to participate in the content assist process. However, it would be better if bnd did not do this indexing directly on the main thread, for these reasons. Better still would be if the index were pre-calculated by a background thread ahead of time instead of on-demand, but I'm not sure what this does for the local space requirements.

**Comment by pkriens** [comment]
Did it really freeze? I thought I'd seen that quickfixes were executed in a job?

**Comment by gamerson** [comment]
Yes, I had to `kill -9` eclipse after 5-10min.  

"content assist proposals" seem also reach into "quick fix" proposals and those are executed on UI thread.

Also, I executed the "quick fix" from the "problems" view, and I get the same UI thread hang.  here is stacktrace for UI thread starvation from "problems view > right click a java problem > quick fix".  
https://gist.github.com/gamerson/b960a00d1557664f55bd370b43141b90

However... with the "quick fix" invoked from the Problems View, that will eventually resolve (after about 30 seconds), whereas the "Ctrl + 1 quick fix" invocation from the editor context hangs indefinitely.
